src/eval/mod.rs

```diff
@@ -0,0 +1,85 @@
+use std::convert::TryFrom;
+use std::io::Write;
+
+use crate::ast::{BinOp, UnOp};
+
+pub mod machine;
+mod lang_env;
+
+use machine::Machine;
+use lang_env::LangEnv;
+
+pub struct OverflowMachine<W: Write> {
+    env: LangEnv<W>,
+}
+
+impl<W: Write> Machine<W> for OverflowMachine<W> {
+    fn with_env(env: LangEnv<W>) -> Self {
+        OverflowMachine { env }
+    }
+
+    fn lang_env(&mut self) -> &mut LangEnv<W> {
+        &mut self.env
+    }
+}
+
+pub struct CheckedMachine<W: Write> {
+    env: LangEnv<W>,
+}
+
+impl<W: Write> Machine<W> for CheckedMachine<W> {
+    fn with_env(env: LangEnv<W>) -> Self {
+        CheckedMachine { env }
+    }
+
+    fn lang_env(&mut self) -> &mut LangEnv<W> {
+        &mut self.env
+    }
+
+    fn native_bin_op(op: BinOp, n1: i64, n2: i64) -> i64 {
+        use BinOp::*;
+
+        let result = match op {
+            Add => n1.checked_add(n2),
+            Sub => n1.checked_sub(n2),
+            Mul => n1.checked_mul(n2),
+            Div => n1.checked_div(n2),
+            Rem => n1.checked_rem(n2),
+            Lt => Some((n1 < n2).into()),
+            Lte => Some((n1 <= n2).into()),
+            Gt => Some((n1 > n2).into()),
+            Gte => Some((n1 >= n2).into()),
+            Eq => Some((n1 == n2).into()),
+            Neq => Some((n1 != n2).into()),
+            BitAnd | And => Some(n1 & n2),
+            BitOr | Or => Some(n1 | n2),
+            BitXor => Some(n1 ^ n2),
+            Shr => u32::try_from(n2).ok().and_then(|n2| n1.checked_shr(n2)),
+            Shl => u32::try_from(n2).ok().and_then(|n2| n1.checked_shl(n2)),
```

Comment:
Can you check if using the `overflowing_XXX` returning `(value, overflow_occurred)` methods result in more readable code? Maybe it doesn't but if it does we can skip the creation of the `Option`'s.

Reply:
I think we could but then how should we handle `Shr` and `Shl`? both `checked_sh*` and `overflowing_sh*` require an `u32`.

Reply:
I like it